### PR TITLE
Data collator for token classification pads labels column when receives pytorch tensors

### PR DESCRIPTION
**Pull request by markovalexander**
# What does this PR do?

Fixes DataCollatorForTokenClassification fails when trying to collate examples with "labels" column of pytorch tensors. 

I faced this issue yesterday and checked that it can be solved easily, so I did not open any Issue on github. 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by markovalexander** [comment]
@sgugger

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -305,20 +305,30 @@ def torch_call(self, features):
 
         label_name = "label" if "label" in features[0].keys() else "labels"
         labels = [feature[label_name] for feature in features] if label_name in features[0].keys() else None
+
+        if labels is not None:
+            no_label_features = [{k: v for k, v in feature.items() if k != label_name} for feature in features]
+        else:
+            no_label_features = features
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:312]: I don't think you need the test here,
```suggestion
        no_label_features = [{k: v for k, v in feature.items() if k != label_name} for feature in features]
```
will always work.

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -305,20 +305,30 @@ def torch_call(self, features):
...
+
         batch = self.tokenizer.pad(
-            features,
+            no_label_features,
             padding=self.padding,
             max_length=self.max_length,
             pad_to_multiple_of=self.pad_to_multiple_of,
-            # Conversion to tensors will fail if we have labels as they are not of the same length yet.
             return_tensors="pt" if labels is None else None,
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:319]:
```suggestion
            return_tensors="pt",
```
Can always be this now.

File: src/transformers/data/data_collator.py [hunk 3]
```diff
@@ -305,20 +305,30 @@ def torch_call(self, features):
...
 
         if labels is None:
             return batch
 
-        sequence_length = torch.tensor(batch["input_ids"]).shape[1]
+        sequence_length = (
+            torch.tensor(batch["input_ids"]).shape[1]
+            if not isinstance(batch["input_ids"], torch.Tensor)
+            else batch["input_ids"].shape[1]
+        )
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:329]: This won't be necessary since with the suggestion above, `batch["input_ids"]` will always be a tensor.

File: src/transformers/data/data_collator.py [hunk 4]
```diff
@@ -328,7 +338,9 @@ def torch_call(self, features):
                 [self.label_pad_token_id] * (sequence_length - len(label)) + list(label) for label in labels
             ]
 
-        batch = {k: torch.tensor(v, dtype=torch.int64) for k, v in batch.items()}
+        batch = {
+            k: (torch.tensor(v, dtype=torch.int64) if not isinstance(v, torch.Tensor) else v) for k, v in batch.items()
+        }
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:343]: Only the `label_name` key needs to be converted here, so maybe do this in the test before directly?

File: src/transformers/data/data_collator.py [hunk 5]
```diff
@@ -328,7 +338,9 @@ def torch_call(self, features):
                 [self.label_pad_token_id] * (sequence_length - len(label)) + list(label) for label in labels
             ]
```
Comment (sgugger) [inline src/transformers/data/data_collator.py:339]: If the `labels` are tensors, `list(label)` won't convert the `label` to a list, you need to call `tolist()`.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. I have a couple of comments.

**Review by sgugger — APPROVED** [review]
Thanks for iterating!
